### PR TITLE
#6797: Unique interval classification for styles

### PR DESCRIPTION
**Pull request by dsuren1**
## Description
This PR adds commits to implement unique interval method classification for style editor

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Feature

## Issue

**What is the current behavior?**
#6797 

**What is the new behavior?**
- Vector styles can be modified with new classification method **Unique interval**

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/api/StyleEditor.js [hunk 1]
```diff
@@ -194,46 +222,65 @@ export function classificationVector({
...
+            paramSLDService = {
+                intervals: params.intervals,
+                method,
+                attribute: params.attribute,
+                reverse: params.reverse,
+                ...rampParams,
+                ...(_tempClasses && {tempIntervals: intervalsForUnique})
+            };
+            paramSLDService = parseForUniqueInterval(params, paramSLDService);
+            return axios.get(SLDService.getStyleMetadataService(layer, paramSLDService))
```
[the remark this answers is not in the document]
Comment (dsuren1) [reply web/client/api/StyleEditor.js:265]: In that case, can we generate colors for all methods in the client side itself, so we can avoid two different blocks of code?
```
return updateRules(ruleId, rules, (rule) => ({
                ...rule,
                ...values,
                classification: updateRulesWithColor(data, params),
                errorId: undefined
            }));
```

File: web/client/components/style/ThemaClassesEditor.jsx [hunk 2]
```diff
@@ -40,17 +40,23 @@ class ThemaClassesEditor extends React.Component {
                     format="hex"
                     onChangeColor={(color) => this.updateColor(index, color)}
                 />
-                <NumberPicker
+                { classItem.equal ? <NumberPicker
```
[the remark this answers is not in the document]
Comment (dsuren1) [reply web/client/components/style/ThemaClassesEditor.jsx:43]: Actually this prop `equal` is generated in rules creation for unique interval value, just like min & max for other methods. Anyway I will change this to `unique`, if that would be clear enough

File: web/client/translations/data.en-US.json [hunk 3]
```diff
@@ -2225,6 +2227,7 @@
             "emptyRuleEditor": "There are not rules for this style. Add a new rule by clicking on the desidered symbolizer type in the toolbar",
             "classificationError": "Service unavailable. Currently it is not possible to create a classification",
             "classificationRasterError": "Service unavailable. Currently it is not possible to create a raster classification",
+            "classificationUniqueIntervalError": "Style classes produced by the classification is greater than the intervals for unique specified",
```
Comment (allyoucanmap) [inline web/client/translations/data.en-US.json:2230]: Maximum intervals limits exceeded! You are trying to generate an unique interval classification with an attribute that contains more than {intervalsForUnique} different classes. Unique Interval classification should be applied to attributes with a limited set of intervals to allow a better readability of the style. You could try different method for the selected attribute.

We could use the msgParams for {intervalsForUnique}

please add translation also for other languages

File: web/client/translations/data.it-IT.json [hunk 4]
```diff
@@ -2214,6 +2216,7 @@
             "emptyRuleEditor": "Non ci sono regole per questo stile. Aggiungi una nuova regola facendo clic sul tipo di stile desiderato nella barra degli strumenti",
             "classificationError": "Servizio non disponibile. Attualmente non è possibile creare una classificazione",
             "classificationRasterError": "Servizio non disponibile. Attualmente non è possibile creare una classificazione raster",
+            "classificationUniqueIntervalError": "Le classi di stile prodotte dalla classificazione sono maggiori degli intervalli per univoco specificato",
```
Comment (allyoucanmap) [inline web/client/translations/data.it-IT.json:2219]: Superato il limite massimo di intervalli! Stai tentando di generare una classificazione ad intervallo univoco con un attributo che contiene più di {intervalsForUnique} classi diverse. La classificazione ad intervallo univoco dovrebbe essere applicata agli attributi con un insieme limitato di intervalli per consentire una migliore leggibilità dello stile. Puoi provare un metodo diverso per l'attributo selezionato.

**Review by allyoucanmap — CHANGES_REQUESTED** [review]
There is also this behaviour that should be fixed, steps:

- select Unique Interval method
- select a string attribute
- change method to one that does not support string

_Expected result_:

We have two options:
- 1 disable method dropdown if the selected attribute is a string
- 2 or make a clear error message

_Current result_:

- classification error
string

File: web/client/translations/data.it-IT.json (hunk 4 above)
Comment (tdipisa) [reply web/client/translations/data.it-IT.json:2219]: @offtherailz ok for me.

Comment (offtherailz) [reply web/client/translations/data.it-IT.json:2219]: wrong mention?

**Review by allyoucanmap — CHANGES_REQUESTED** [review]
Some issues found while testing:
- [x] I was not able to apply the cfg parameter intervalsForUnique to test the message error. Probably this value should not be saved in the style object
- [x] If you try to add a text symbolizer the application becomes unresponsive

@tdipisa apart from the issues above the PR follows the requested changes. Now the method select input is disabled if the attribute does not support others method types (eg strings supports only unique interval). After testing this new behaviour it seems a bit misleading and not so clear and probably we will need something else in the ui to explain the reason of why the method is disabled.
see here
string-mss

**Comment by tdipisa** [comment]
@allyoucanmap 
>After testing this new behaviour it seems a bit misleading and not so clear and probably we will need something else in the ui to explain the reason of why the method is disabled.
see here

Is probably enough to reset the method drop down in case the selected method is not available for the selected attribute. In the DD list the method will remain obviously disbled and the user cannot select it.

**Comment by simboss** [comment]
@tdipisa @allyoucanmap @dsuren1 I need this like now, I can't wait any longer.
This PR has been open 11 days ago!

**Review by allyoucanmap — APPROVED** [review]

**Comment by allyoucanmap** [comment]
@tdipisa @dsuren1 as soon as this PR as been tested it needs to be backported to https://github.com/geosolutions-it/MapStore2/tree/geonode-mapstore-client-2.1.x if we need it on the stable branch of GeoNode

**Comment by tdipisa** [comment]
@ElenaGallo please test this in DEV as soon as deployed and let us know if we can proceed with the backport to geonode-mapstore-client-2.1.x branch

**Comment by ElenaGallo** [comment]
Test done, @tdipisa PR can be backported to geonode-mapstore-client-2.1.x branch

**Pull request merged**